drivers/i2c/i2c_ll_stm32_v2.c

```diff
@@ -248,7 +250,7 @@ int stm32_i2c_msg_read(struct device *dev, struct i2c_msg *msg,
 		len--;
 	}
 
-	msg_done(dev, flags | msg->flags);
+	msg_done(dev, flags);
```

Comment:
The flag I2C_MSG_RESTART inside this driver means 
>this message is followed by a restart condition, so don't issue a stop condition at the end of this message.

This differs from the meaning of the I2C_MSG_RESTART in the general Zephyr I2C API
>I2C_MSG_RESTART (1 << 2)
RESTART I2C transaction for this message.

To fix this both the stm32-i2c-v1 and v2 driver need a rework, which is, why I opened Issue #4459.
 
the variable flags is set to I2C_MSG_RESTART in i2c_ll_stm32.c if the current message is followed by one that has a different direction (R/W). So passing it into msg_done function leads to no stop condition beeing issued between those two messages, like intended.

msg->flags is controlled by the app/user, having the API meaning of the I2C_MSG_RESTART flag in mind. Passing it into msg_done would lead to no stop condition being issued after this message, while the user/app intended to start this message with a restart condition.

So until the meanings of I2C_MSG_RESTART in stm32 drivers and API are harmonized we should not use msg->flags inside this driver. 

Daniel Wagenknecht

Reply:
just saw, that this applys to the interrrupt driven driver as well (line 182/184). Changed it with the latest update of this pull request (and rebased onto latest arm branch).